### PR TITLE
Updated code for latest single zarr archive

### PR DESCRIPTION
**Pull request by sadamov**
This PR updates the dataloader and the model to handle both forcing features and additional state features. This was in preparation for the upcoming full retraining of the cosmo-emulator.
- Forcing features (vary temporally but are not predicted)  are added using three new channels - radiation fluxes
- Static features (constant in time not predicted) are updated - `HFL` is not required when `FI` and `HSURF` are available.
- Additional state variables (vary temporally and are predicted) were added as available in the latest zarr archive.

Code was tested by myself, requires one review to spot obvious coding mistakes.

### CODE REVIEW

File: slurm_train.sh [hunk 1]
```diff
@@ -1,37 +1,49 @@
...
 fi
 
 ulimit -c 0
-export OMP_NUM_THREADS=16
 
-# Run the script with torchrun
-srun -ul python train_model.py --dataset "cosmo" --val_interval 5 \
-    --epochs 10 --n_workers 6 --batch_size 8 --subset_ds 1
+echo "Training model"
+if [ "$MODEL" = "hi_lam" ]; then
```
Comment (twicki) [inline slurm_train.sh:45]: I think it is a bity ugly to not just be able to have the entire thing duplicated for an extra arg. But it would work

**Review by twicki — CHANGES_REQUESTED** [review]

File: environment.yml [hunk 2]
```diff
@@ -19,7 +19,6 @@ dependencies:
   - pyproj
   - pyprojroot
   - pytorch=2.2.2=py3.12_cuda11.8_cudnn8.7.0_0
-  - pytorch-cuda=11.8
```
[the remark this answers is not in the document]
Comment (sadamov) [reply environment.yml:22]: without pin on some machines the cpu pytorch variant would be installed, as seen in your last review. Then the default slurm submission scripts fail.

File: neural_lam/constants.py [hunk 3]
```diff
@@ -24,103 +21,144 @@
...
     "2-meter Temperature",
+    "Total Precipitation",
     "10-meter Zonal wind speed",
     "10-meter Meridional wind speed",
 ]
 
+
 # Short names
 PARAM_NAMES_SHORT = [
+    "PP",
```
[the remark this answers is not in the document]
Comment (sadamov) [reply neural_lam/constants.py:53]: There was confusion around this. after discussion with @cosunae it was agreed that:

- model levels: have a fixed height -> constant geopotential in time -> pressure is being modelled
- pressure levels: have a fixed pressure -> geopotential is being modelled

File: slurm_train.sh (hunk 1 above)
Comment (sadamov) [reply slurm_train.sh:43]: yes. so I created these slurm scripts for developers first. and to switch between models, the checkpoint, the graph and the submission script must align. To make it easier for devs, I made it a simple switch at the top of the script that is hard to miss. I suggest to leave this for now for train, eval and predict. For MLOps (e.g. `slurm_offline.py` a more "operational" setup should be chosen.

File: neural_lam/models/ar_model.py [hunk 4]
```diff
@@ -331,7 +335,7 @@ def common_step(self, batch):
             where index 0 corresponds to index 1 of init_states
         """
         init_states, target_states, batch_time = batch[:3]
-        forcing_features = batch[4] if len(batch) > 3 else None
+        forcing_features = batch[3] if len(batch) > 3 else None
```
[the remark this answers is not in the document]
Comment (sadamov) [reply neural_lam/models/ar_model.py:334]: No, it should be "if there are more than 3 elements in batch -> the fourth (at index 3) is forcing)". I think that is how python counts

File: create_parameter_weights.py [hunk 5]
```diff
@@ -89,9 +89,11 @@ def main():
     for batch_data in tqdm(loader):
         if constants.GRID_FORCING_DIM > 0:
             init_batch, target_batch, _, forcing_batch = batch_data
-            flux_batch = forcing_batch[:, :, :, 0]  # Flux is first index
-            flux_means.append(torch.mean(flux_batch))  # (,)
-            flux_squares.append(torch.mean(flux_batch**2))  # (,)
+            flux_batch = forcing_batch[
+                :, :, :, :3
+            ]  # fluxes are first 3 features
+            flux_means.append(torch.mean(flux_batch, dim=(1, 2, 3)))  # (,)
```
[the remark this answers is not in the document]
Comment (sadamov) [reply create_parameter_weights.py:95]: I commited a revamp of this script just now. There were bugs in the code and I deviated too far from upstream. Comments still present in current scripts are mostly from upstream. I suggest to remain aligned for the time being.

File: create_parameter_weights.py [hunk 6]
```diff
@@ -89,9 +89,11 @@ def main():
     for batch_data in tqdm(loader):
         if constants.GRID_FORCING_DIM > 0:
             init_batch, target_batch, _, forcing_batch = batch_data
-            flux_batch = forcing_batch[:, :, :, 0]  # Flux is first index
-            flux_means.append(torch.mean(flux_batch))  # (,)
-            flux_squares.append(torch.mean(flux_batch**2))  # (,)
+            flux_batch = forcing_batch[
```
[the remark this answers is not in the document]
Comment (sadamov) [reply create_parameter_weights.py:92]: That is a great point and I think some of the absolut indices are problematic. My logic so far was to stay close to upstream and introduce flexibility where needed. In terms of forcings, this PR should get us aligned with upstream. I suggest to make PRs directly to upstream for fundamental changes like updating the dataflow and tensor shape of forcings.

**Comment by sadamov** [comment]
Can I merge this PR? Zarr archive is ready and I need to start full retraining before weekend.

**Pull request merged**